service/history/shard/context_impl.go

```diff
@@ -949,34 +967,44 @@ func (s *ContextImpl) DeleteWorkflowExecution(
 		}
 	}
 
-	// Don't acquire shard lock if all stages that require lock are already processed.
+	validateCtxAndShardState := func() (context.Context, context.CancelFunc, error) {
```

Comment:
I miss the meaning of this one. The shard could be in a error state after this returns a context?

Reply:
Yes, this check is just best effort and mainly for making sure context has enough time left. 

In this PR, shard lock is released when making all shard I/O request. And yeah, when those calls are made, shard can be in an error state. But that's fine, if the request really requires shard ownership, it will be conditioned on rangeID (to handle shard stealing). In this DeleteWorkflow case, we currently put those calls (step 2 and 3) in shard lock just for concurrency control purpose.